File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/dashboards/widgets/home/LastActiveProjectsWidget.kt
package net.nemerosa.ontrack.model.dashboards.widgets.home

import net.nemerosa.ontrack.json.asJson
import net.nemerosa.ontrack.model.dashboards.WidgetInstance
import net.nemerosa.ontrack.model.dashboards.widgets.AbstractWidget
import net.nemerosa.ontrack.model.dashboards.widgets.WidgetConfig
import org.springframework.stereotype.Component

@Component
class LastActiveProjectsWidget : AbstractWidget<LastActiveProjectsWidget.LastActiveProjectsWidgetConfig>(
    key = "home/LastActiveProjects",
    name = "Last active projects"
) {
    data class LastActiveProjectsWidgetConfig(
        val count: Int = 10,
    ) : WidgetConfig

    fun toInstance(count: Int = 10) = WidgetInstance(
        key,
        LastActiveProjectsWidgetConfig(count).asJson()
    )
}

File: ontrack-web-core/components/widgets/home/LastActiveProjectsWidget.js
import {useContext} from "react";
import {useEventForRefresh} from "@components/common/EventsContext";
import {gql} from "graphql-request";
import {gqlDecorationFragment} from "@components/services/fragments";
import {DashboardWidgetCellContext} from "@components/dashboards/DashboardWidgetCellContextProvider";
import PaddedContent from "@components/common/PaddedContent";
import SimpleProjectList from "@components/projects/SimpleProjectList";
import {gqlProjectContentFragment} from "@components/projects/ProjectGraphQLFragments";
import {useQuery} from "@components/services/GraphQL";

export default function LastActiveProjectsWidget({count}) {

    const projectsRefreshCount = useEventForRefresh("project.created")
    const favouritesRefresh = useEventForRefresh("project.favourite")

    const {setTitle} = useContext(DashboardWidgetCellContext)
    setTitle(`Last ${count} active projects`)

    const {data: projects} = useQuery(
        gql`
            query LastActiveProjects($count: Int! = 10) {
                lastActiveProjects(count: $count) {
                    ...ProjectContent
                    favourite
                    decorations {
                        ...decorationContent
                    }
                }
            }

            ${gqlProjectContentFragment}
            ${gqlDecorationFragment}
        `,
        {
            variables: {count},
            initialData: [],
            deps: [count, projectsRefreshCount, favouritesRefresh],
            dataFn: data => data.lastActiveProjects,
        }
    )

    return (
        <PaddedContent>
            <SimpleProjectList
                projects={projects}
                emptyText={
                    <>
                        No project has been created in Ontrack yet.
                        You can start <a
                        href="https://static.nemerosa.net/ontrack/release/latest/docs/doc/index.html#feeding">feeding
                        information</a> in Ontrack
                        automatically from your CI engine, using its API or other means.
                    </>
                }
            />
        </PaddedContent>
    )
}